Doc/enterprise#334 - Cluster mode documentation (#937)
Avaiga/taipy-doc: docs/manuals/core/config/job-config.md

> The `JobConfig^` allows the developer to configure the Taipy behavior for job executions. There are three modes

Reviewer: ```suggestion
`JobConfig^` allows the developer to configure the Taipy behavior for job executions. There are three modes
```
or
```suggestion
The `JobConfig^` class allows the developer to configure the Taipy behavior for job executions. There are three modes
```

> available in Taipy: the *standalone* mode, the *development* mode, and the *cluster* mode

Reviewer: ```suggestion
available in Taipy: *standalone*, *development*, and *cluster*.
```

With the *standalone* mode, a `Job^` runs in its own execution context, in an asynchronous manner.
At the submission, the job is queued. It is dequeued in a different thread and sent to a dedicated process to be
executed.

> Note that with the *standalone* mode, the submit method is not blocking and returns after the job is queued.

Reviewer: ```suggestion
Note that with the *standalone* mode, the `(Scenario.)submit()^` method is not blocking and returns after the job is queued.
```

    max_nb_of_workers = "8:int"
    ```

# Cluster mode

!!! warning "Available in Taipy Enterprise edition"

    This section is relevant only to the Enterprise edition of Taipy.

> With the *cluster* mode, a `Job^` runs in its own environment (a worker environment), which can be
> local or on a remote machine, in an asynchronous manner. At the submission, the job is queued in a
> RabbitMQ queue. When there is a worker available, the job is dequeued and sent to the worker
> environment to be executed.

Reviewer 2: Trying to add some motivation. What do you think?

```suggestion
The *cluster* mode is designed to make a Taipy application execute the jobs on a remote cluster of distributed workers. The number of machines available in the cluster is unlimited, so the application becomes fully scalable horizontally. It can also be adjusted dynamically depending on some load metric.
With the *cluster* mode, a `Job^` asynchronously runs in its own environment. At the submission, the job is queued in a RabbitMQ queue. When there is a worker available, the job is dequeued and sent to the worker environment to be executed.
```

Author: I realigned the motivation sentence and push a new commit. I think this is a great addition to better explain this mode

Note that with the *cluster* mode, the submit method is not blocking and returns after the job is queued.
It means the submit method can return before the job finishes or even before it is dequeued. Please refer to
the [submit entity](../entities/orchestrating-and-job-execution.md#submit-a-scenario-sequence-or-task) section
on how to submit jobs.

You can configure the *cluster* mode with the following config:

=== "Python configuration"

    ```python linenums="1"
    from taipy import Config

    Config.configure_job_executions(mode="cluster")
    ```

=== "TOML configuration"

    ```python linenums="1"
    from taipy import Config

    Config.load("config.toml")
    ```

    ```toml linenums="1" title="config.toml"
    [JOB]
    mode = "cluster"
    ```

To set up a worker environment, some requirements must be met:

> - The worker environment must have access to the same Taipy application as the main environment.
> - The worker environment must have the same Python environment as the main environment.
> - The worker environment must have access to the same RabbitMQ server as the main environment.
> - The worker environment must have access to the data of the data nodes as the main environment.
>   This includes:
>   - The shared drive(s) containing the data for file-based data nodes.
>   - The database server(s) containing the data for database-based data nodes.

Reviewer 2: Reformulation. What do you think?

```suggestion
- The main application and the workers must have the same Python environment.
- The main application and the workers must have access to the same RabbitMQ server.
- The main application must be configured to use a Mongo repository, and the workers must have access to it as well.
- The main application and the workers must have access to the data referenced by the the data nodes. This includes:
    - The shared drive(s) containing the data for file-based data nodes.
    - The database server(s) containing the data for database-based data nodes.
```

Author: The idea behind my proposal is to explain for the user how to set up 1 worker. I call each worker as "worker environment" because the worker can be remote or local, it just needs to be a separate environment.

About the "Mongo repository" requirement, how about "must have access to the entities"?
Mongo is nice to have, but not strictly.

Reviewer 2: 1. **About worker environment:** For me, the whole feature makes sense only on remote machines. It does not make sense to me to run a local worker. We have the standalone mode for that. This is only useful for us as contributors, but not really for our users. What do you think?

2. **About mongo:** I thought a `mongo` repository was mandatory to run in `cluster` mode. Is it possible to run a remote worker on a filesystem repository, for instance? If yes, I agree with your proposal.

Author: 1. That makes sense. Let me rephrase this
2. Yes, the user can use cluster mode as long as all machines have access to the same entity repository. For FS and SQLite, it can be a shared folder or file. For Mongo, it's more convinient

    max_nb_of_workers = "8:int"
    ```

# Cluster mode

!!! warning "Available in Taipy Enterprise edition"

    This section is relevant only to the Enterprise edition of Taipy.

> The *cluster* mode is designed to make a Taipy application executes jobs on a remote cluster of

Reviewer: ```suggestion
The *cluster* mode is designed to make Taipy applications execute jobs on a remote cluster of
```

distributed workers. The number of machines available in the cluster is unlimited, so the
application becomes fully scalable horizontally. It can also be adjusted dynamically depending on

> some load metric.

Reviewer: ```suggestion
some load metrics.
```

> With the *cluster* mode, a `Job^` asynchronously runs in its own environment. At the submission,

Reviewer: ```suggestion
With the *cluster* mode, a `Job^` asynchronously runs in its own environment. When submitted,
```

> the job is queued in a RabbitMQ queue. When there is a worker available, the job is dequeued and

Reviewer: Do we need to mention the technology (RabbitMQ) at this point?

Author: We probably not. What do you think @jrobinAV?

Reviewer 2: I would say yes. 
We rely on RabbitMQ. Users need to know it to start it and manage it.

sent to the worker environment to be executed.

> Note that with the *cluster* mode, the submit method is not blocking and returns after the job is queued.

Reviewer: ```suggestion
Note that with the *cluster* mode, the `(Scenario.)submit()^` method is not blocking and returns after the job is queued.
```

> It means the submit method can return before the job finishes or even before it is dequeued. Please refer to

Reviewer: ```suggestion
This means the `submit()` method can return before the job finishes or even before it is dequeued. Please refer to
```

the [submit entity](../entities/orchestrating-and-job-execution.md#submit-a-scenario-sequence-or-task) section
on how to submit jobs.

You can configure the *cluster* mode with the following config:

=== "Python configuration"

    ```python linenums="1"
    from taipy import Config

    Config.configure_job_executions(mode="cluster")
    ```

=== "TOML configuration"

    ```python linenums="1"
    from taipy import Config

    Config.load("config.toml")
    ```

    ```toml linenums="1" title="config.toml"
    [JOB]
    mode = "cluster"
    ```

To set up a worker environment, some requirements must be met:

- The worker environment must have access to the same Taipy application as the main environment.
- The worker environment must have the same Python environment as the main environment.

> - The worker environment must have access to the same RabbitMQ server as the main environment.

Reviewer: Do we need to expose the techno here?

Author: Most is necessary I think. However, for the "RabbitMQ", it may be too specific. What do you think @jrobinAV?

Reviewer 2: That is a prerequisite. We rely on RabbitMQ for this feature. I would keep it here.
